zip-0221.rst

```diff
@@ -0,0 +1,585 @@
+        ...
+    }
+
+
+Rationale
+-----------
+
+Tree nodes
+~~~~~~~~~~~
+
+Nodes in the commitment tree are canonical and immutable. They are cheap to generate, as (with the exception of ``nShieldedTxCount``) all metadata is already generated during block construction and/or checked during block validation. Nodes are relatively compact in memory. Approximately 140,000 blocks have elapsed since Sapling activation. Assuming a 164 byte commitment to each of these, we would have generated approximately 24 MB of additional storage cost for the set of leaf nodes (and an additional ~24 MB for storage of intermediate nodes).
+
+``hashSubtreeCommitment`` forms the strucuture of the commitment tree. Other metadata commitments were chosen to serve specific purposes. Variable-length commitments are placed last, so that most metadata in a node can be directly indexed. We considered using fixed-length commitments here, but opted for variable-length, in order to marginally reduce the memory requirements for managing and updating the commitment trees.
+
+In leaf nodes, some information is repeated. We chose to do this so that leaf nodes could be treated identically to internal and root nodes for all algorithms and (de)serializers. Leaf nodes are easily identifiable, as they will show proof of work in the ``hashSubtreeCommitment`` field, and their block range (calculated as ``nLatestHeight - nEarliestHeight + 1``) will be precisely 1. For the same reason, we change the semantics of ``hashSubtreeCommitment`` in leaf nodes to commit
+
+Personalized BLAKE2b-256 was selected to match existing ZCash conventions. Adding the consensus branch ID to the hash personalization string ensures that valid nodes from one branch cannot be used to make false statements about parallel consensus branches.
+
+FlyClient Requirements and Recommendations
+===============================================
+These commitments enable FlyClient in the variable-difficulty model. Specifically, they allow light clients to reason about application of the difficulty adjustment algorithm over a range of blocks. They were chosen via discussion with an author of the FlyClient paper.
+
+- ``nEarliestTimestamp``
+- ``nLatestTimestamp``
+- ``nEarliestTarget``
+- ``nLatestTarget``
+- ``nEarliestHeight``
+- ``nLatestHeight``
+- ``nSubTreeTotalWork``
+
+Non-FlyClient Commitments
+==========================
+
+Additional metadata commitments were chosen primarily to improve light client security guarantees. We specified commitments where we could see an obvious security benefit, but there may be other useful metadata that we missed. We're interested in feedback and suggestions from the implementers of the current light client.
+
+We considered adding a commitment to the nullifier vector at each block. We would appreciate comments from light client teams on the utility of this commitment, as well as the proper serialization and commitment format for the nullifier vector.
```

Comment:
At some point we will want to commit to the nullifier set for Sapling, but the commitment structure is not yet defined. It would be useful to describe how the MMR node structures should be altered across network upgrades (which would be the boundaries at which we would be adding or removing commitments); in particular, the behaviour of, or limitations on, MMR trees that cross a node-structure-altering network upgrade activation height should be specified.

Relatedly, the transaction malleability fix that is being considered for NU4 would itself add a new per-block commitment (functionally equivalent to the witness tree root for SegWit). The commitments currently included are either block properties, or the roots of global (not per-block) accumulators; I take it that means per-block commitments should not be part of MMR nodes, and would instead require a block header format change.

Reply:
> in particular, the behaviour of, or limitations on, MMR trees that cross a node-structure-altering network upgrade activation height should be specified.

This is implicitly addressed earlier. There should be no such tree.

```The leaves of the MMR at block ``B_n`` are hash commitments to the header data and metadata of each previous block ``B_x, ..., B_(n-1)``, where ``x`` is the block height of the most recent network upgrade.```

Committing to blocks that were created before a (hard fork) network upgrade adds no meaningful benefit to light clients, as they must be upgraded to explicitly encode info about the NU anyway.

This also ensures that node structure can be rewritten entirely at each NU if desired. 

> The commitments currently included are either block properties, or the roots of global (not per-block) accumulators; I take it that means per-block commitments should not be part of MMR nodes, and would instead require a block header format change.

The block height is a per-block commitment that gets preserved in the tree structure. You could fold in any per-block commitment to the existing MMR tree by specifying that parent nodes keep the right child node's commitment. That would allow validation of the committed information only the preimage of the tree root commitment

Whether it's valuable to make a separate header commitment is another question.

Reply:
> This is implicitly addressed earlier.

Excellent! It would be good to make this more visible / explicit in the ZIP (I read right past it without noticing the implications of the end of the sentence), and/or put the above comments in as rationale.

> This also ensures that node structure can be rewritten entirely at each NU if desired.

:+1: (obviously it's easier from an implementation perspective if the rewrite is minimal, but good that the ZIP makes it possible).

> Whether it's valuable to make a separate header commitment is another question.

This is something I'm undecided on. On the one hand, the MMR root hash is fine for binding these extra commitments to the header, and fewer commitments in the header means less parsing work for the ecosystem areas that touch the header (but not less work, because they all still need to know how to fill the MMR root hash). On the other hand, using and verifying a particular commitment within an MMR leaf requires all the MMR / FlyClient data structures and logic (e.g. making this change places a hard requirement on light clients, `lightwalletd` etc. to implement it all if they want to continue to verify their Sapling tree roots). If a commitment will be regularly used by everyone in the network (rather than subsets of participants) then it is already logically part of the core header, and it's simpler and more efficient for the commitment to be part of the actual serialized header.

Reply:
> If a commitment will be regularly used by everyone in the network (rather than subsets of participants) then it is already logically part of the core header, and it's simpler and more efficient for the commitment to be part of the actual serialized header.

Zcash headers are already >1Kb, so it doesn't seem like adding additional commitments will add significant additional overhead to anyone. The question, I guess, becomes how do you sort proposed new commitments into "header" or "tree commitment"

I think it's a bit more complex than looking at how many people use it, as there's additional utility to range commitments for a lot of information (e.g. difficulty change information). Also, light clients are inherently a subset of the network. Full nodes likely won't use a nullifier set commitment as they already have one